src/apdu.rs

```diff
@@ -0,0 +1,200 @@
+    sign_data.extend(challenge);
+    sign_data.extend(application);
+    sign_data.push(key_handle.len() as u8);
+    sign_data.extend(key_handle);
+
+    let flags = U2F_CHECK_IS_REGISTERED;
+    let (_, status) = dev.send_apdu(U2F_AUTHENTICATE, flags, &sign_data)?;
+    Ok(status == SW_CONDITIONS_NOT_SATISFIED)
+}
+
+pub fn apdu_is_v2_device<T>(dev: &mut T) -> io::Result<bool>
+where
+    T: APDUDevice,
+{
+    let (data, status) = dev.send_apdu(U2F_VERSION, 0x00, &[])?;
+    let actual = CString::new(data)?;
+    let expected = CString::new("U2F_V2")?;
+    apdu_status_to_result(status, actual == expected)
+}
+
+////////////////////////////////////////////////////////////////////////
+// Error Handling
+////////////////////////////////////////////////////////////////////////
+
+pub fn apdu_status_to_result<T>(status: [u8; 2], val: T) -> io::Result<T> {
+    use self::io::ErrorKind::{InvalidData, InvalidInput};
+
+    match status {
+        SW_NO_ERROR => Ok(val),
+        SW_WRONG_DATA => Err(io::Error::new(InvalidData, "wrong data")),
+        SW_WRONG_LENGTH => Err(io::Error::new(InvalidInput, "wrong length")),
+        SW_CONDITIONS_NOT_SATISFIED => Err(io_err("conditions not satisfied")),
+        _ => Err(io_err(&format!("failed with status {:?}", status))),
+    }
+}
+
+// https://en.wikipedia.org/wiki/Smart_card_application_protocol_data_unit
+// https://fidoalliance.org/specs/fido-u2f-v1.
+// 0-nfc-bt-amendment-20150514/fido-u2f-raw-message-formats.html#u2f-message-framing
+pub struct APDU {}
+
+impl APDU {
+    pub fn serialize_long(ins: u8, p1: u8, data: &[u8]) -> io::Result<Vec<u8>> {
+        let class: u8 = 0x00;
+        if data.len() > 0xffff {
+            return Err(io_err("payload length > 2^16"));
+        }
+
+        // Size of header + data + 2 zero bytes for maximum return size.
+        let mut bytes = vec![0u8; U2FAPDUHEADER_SIZE + data.len() + 2];
+        bytes[0] = class;
+        bytes[1] = ins;
+        bytes[2] = p1;
+        // p2 is always 0, at least, for our requirements.
+        // lc[0] should always be 0.
+        bytes[5] = (data.len() >> 8) as u8;
+        bytes[6] = data.len() as u8;
+        bytes[7..7 + data.len()].copy_from_slice(data);
+
+        // When sending zero data, the two data length bytes should be omitted.
+        // Luckily, all later bytes are zero, so we can just truncate.
+        if data.is_empty() {
+            bytes.truncate(bytes.len() - 2);
+        }
+
+        Ok(bytes)
+    }
+
+    // This will be used by future NFC code
+    #[allow(dead_code)]
+    pub fn serialize_short(ins: u8, p1: u8, data: &[u8]) -> io::Result<Vec<u8>> {
+        let class: u8 = 0x00;
+        if data.len() > 0xff {
+            return Err(io_err("payload length > 2^8"));
+        }
+
+        let mut size = 5; // class, ins, p1, p2, response size field
+        if !data.is_empty() {
+            size += 1 + data.len(); // data size field and data itself
```

Comment:
This needs a bit more description, because it appears that the data-size field is `bytes[4]`, accounted for in the initial value of `5`. This is actually because of the overflow zeroes?

Reply:
This is already correct: when there is no command data payload (Ne=0), you omit the command length bytes (Lc).